BananaPi BPI-F3: `Update kernel to 6.6.y and other fixups`

pyavitz opened the pull request:
Kernel source: https://github.com/jmontleon/linux-bianbu.git
I've been using it now for a couple weeks and works well.

U-Boot source: https://gitee.com/bianbu-linux/uboot-2022.10.git
OpenSBI source: https://gitee.com/bianbu-linux/opensbi.git
TAG="v2.0.4"

Updated config/sources/families/spacemit.conf
Made the needed adjustments and simplified the write_uboot_platform function.

~Note: Bluetooth is still b0rked~

Note: `rfkill unblock all`.
```
patrick@bananapif3:~$ dmesg | grep hci0
[   10.559798] Bluetooth: hci0: RTL: examining hci_ver=0b hci_rev=000b lmp_ver=0b 
lmp_subver=8852
[   10.563485] Bluetooth: hci0: RTL: rom_version status=0 version=1
[   10.563518] Bluetooth: hci0: RTL: loading rtl_bt/rtl8852bs_fw.bin
[   10.565898] Bluetooth: hci0: RTL: loading rtl_bt/rtl8852bs_config.bin
[   10.587947] Bluetooth: hci0: RTL: cfg_sz 33, total sz 46430
[   11.148639] Bluetooth: hci0: RTL: fw version 0xdec834cf
patrick@bananapif3:~$ lsmod
Module                  Size  Used by
bnep                   28672  2
lz4hc                  12288  0
hci_uart              151552  0
lz4                    12288  0
btqca                  28672  1 hci_uart
btrtl                  36864  1 hci_uart
btbcm                  32768  1 hci_uart
bluetooth             995328  26 btrtl,btqca,hci_uart,btbcm,bnep
ecdh_generic           16384  2 bluetooth
ecc                    49152  1 ecdh_generic
zram                   40960  3
zsmalloc               40960  1 zram
8852bs               7921664  0
cfg80211             1482752  1 8852bs
fuse                  233472  1
nfnetlink              24576  1
ip_tables              28672  0
x_tables               53248  1 ip_tables
uio_pdrv_genirq        16384  0
uio                    28672  1 uio_pdrv_genirq
patrick@bananapif3:~$ bluetoothctl
Waiting to connect to bluetoothd...[bluetooth]# hci0 new_settings: powered bondabl
e ssp br/edr le secure-conn cis-central cis-peripheral 
[bluetooth]# Agent registered
[bluetooth]# [CHG] Controller C0:4B:24:36:6D:C2 Pairable: yes
[bluetooth]# show
Controller C0:4B:24:36:6D:C2 (public)
	Manufacturer: 0x005d (93)
	Version: 0x0b (11)
	Name: bananapif3
	Alias: bananapif3
	Class: 0x00000000 (0)
	Powered: yes
	Discoverable: no
	DiscoverableTimeout: 0x000000b4 (180)
	Pairable: yes
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)
	UUID: A/V Remote Control Target (0000110c-0000-1000-8000-00805f9b34fb)
	UUID: A/V Remote Control        (0000110e-0000-1000-8000-00805f9b34fb)
	UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
	Modalias: usb:v1D6Bp0246d0548
	Discovering: no
	Roles: central
	Roles: peripheral
Advertising Features:
	ActiveInstances: 0x00 (0)
	SupportedInstances: 0x04 (4)
	SupportedIncludes: tx-power
	SupportedIncludes: appearance
	SupportedIncludes: local-name
	SupportedSecondaryChannels: 1M
	SupportedSecondaryChannels: 2M
	SupportedSecondaryChannels: Coded
```

The-going commented:
Build on board BPI-F3
```
leo@bananapif3:~/DATA$ git clone --depth=100 https://github.com/armbian/build -b main armbian
leo@bananapif3:~/DATA$ cd armbian/
leo@bananapif3:~/DATA/armbian$ git remote add -t BPI-F3-007 pyavitz https://github.com/pyavitz/armbian
leo@bananapif3:~/DATA/armbian$ git fetch pyavitz
git checkout -b BPI-F3-007  pyavitz/BPI-F3-007
leo@bananapif3:~/DATA/armbian$ ./compile.sh BOARD=bananapif3 KERNEL_GIT=shallow
```

ended with this error:
```
[🔨]   Collecting pylibfdt (from dtschema==2024.11->-r /home/leo/DATA/armbian/requirements.txt (line 18))
[🔨]     Downloading pylibfdt-1.7.2.tar.gz (61 kB)
[🔨]        ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.6/61.6 kB 584.7 kB/s eta 0:00:00
[🔨]     Installing build dependencies: started
[🔨]     Installing build dependencies: still running...
[🔨]     Installing build dependencies: finished with status 'error'
[🔨]     error: subprocess-exited-with-error
[🔨]     
[🔨]     × pip subprocess to install build dependencies did not run successfully.
[🔨]     │ exit code: 1
[🔨]     ╰─> [91 lines of output]
[🔨]         Collecting setuptools
[🔨]           Using cached setuptools-75.6.0-py3-none-any.whl.metadata (6.7 kB)
[🔨]         Collecting setuptools_scm
[🔨]           Downloading setuptools_scm-8.1.0-py3-none-any.whl.metadata (6.6 kB)
[🔨]         Collecting swig
[🔨]           Downloading swig-4.3.0.tar.gz (25 kB)
[🔨]           Installing build dependencies: started
[🔨]           Installing build dependencies: finished with status 'done'
[🔨]           Getting requirements to build wheel: started
[🔨]           Getting requirements to build wheel: finished with status 'done'
[🔨]           Installing backend dependencies: started
[🔨]           Installing backend dependencies: still running...
[🔨]           Installing backend dependencies: finished with status 'error'
[🔨]           error: subprocess-exited-with-error
[🔨]         
[🔨]           × pip subprocess to install backend dependencies did not run successfully.
[🔨]           │ exit code: 1
[🔨]           ╰─> [63 lines of output]
[🔨]               Collecting setuptools-scm
[🔨]                 Using cached setuptools_scm-8.1.0-py3-none-any.whl.metadata (6.6 kB)
[🔨]               Collecting cmake>=3.15
[🔨]                 Downloading cmake-3.31.2.tar.gz (34 kB)
[🔨]                 Installing build dependencies: started
[🔨]                 Installing build dependencies: finished with status 'done'
[🔨]                 Getting requirements to build wheel: started
[🔨]                 Getting requirements to build wheel: finished with status 'done'
[🔨]                 Preparing metadata (pyproject.toml): started
[🔨]                 Preparing metadata (pyproject.toml): finished with status 'done'
[🔨]               Collecting packaging>=20 (from setuptools-scm)
[🔨]                 Using cached packaging-24.2-py3-none-any.whl.metadata (3.2 kB)
[🔨]               Collecting setuptools (from setuptools-scm)
[🔨]                 Using cached setuptools-75.6.0-py3-none-any.whl.metadata (6.7 kB)
[🔨]               Downloading setuptools_scm-8.1.0-py3-none-any.whl (43 kB)
[🔨]                  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 981.3 kB/s eta 0:00:00
[🔨]               Using cached packaging-24.2-py3-none-any.whl (65 kB)
[🔨]               Using cached setuptools-75.6.0-py3-none-any.whl (1.2 MB)
[🔨]               Building wheels for collected packages: cmake
[🔨]                 Building wheel for cmake (pyproject.toml): started
[🔨]                 Building wheel for cmake (pyproject.toml): still running...
[🔨]                 Building wheel for cmake (pyproject.toml): finished with status 'error'
[🔨]                 error: subprocess-exited-with-error
[🔨]         
[🔨]                 × Building wheel for cmake (pyproject.toml) did not run successfully.
[🔨]                 │ exit code: 1
[🔨]                 ╰─> [30 lines of output]
[🔨]                     ---------------------------------------------
[🔨]                     CMake 3.31.2, Copyright 2000-2024 Kitware, Inc. and Contributors
[🔨]                     C compiler on this system is: cc
[🔨]                     ---------------------------------------------
[🔨]                     Error when bootstrapping CMake:
[🔨]                     Cannot find a C++ compiler that supports both C++11 and the specified C++ flags.
[🔨]                     Please specify one using environment variable CXX.
[🔨]                     The C++ flags are "".
[🔨]                     They can be changed using the environment variable CXXFLAGS.
[🔨]                     See cmake_bootstrap.log for compilers attempted.
[🔨]                     ---------------------------------------------
[🔨]                     Log of errors: /tmp/tmpvdurduom/cmake-3.31.2/Bootstrap.cmk/cmake_bootstrap.log
[🔨]                     ---------------------------------------------
[🔨]                     Traceback (most recent call last):
[🔨]                       File "/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
[🔨]                         main()
[🔨]                       File "/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
[🔨]                         json_out['return_val'] = hook(**hook_input['kwargs'])
[🔨]                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[🔨]                       File "/usr/lib/python3/dist-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 251, in build_wheel
[🔨]                         return _build_backend().build_wheel(wheel_directory, config_settings,
[🔨]                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[🔨]                       File "/tmp/pip-install-oixy72y9/cmake_be0f358760c04b86aafdf564bd3ab675/_build_backend/backend.py", line 156, in build_wheel
[🔨]                         cmake_path = _bootstrap_build(temp_path, config_settings)
[🔨]                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[🔨]                       File "/tmp/pip-install-oixy72y9/cmake_be0f358760c04b86aafdf564bd3ab675/_build_backend/backend.py", line 127, in _bootstrap_build
[🔨]                         subprocess.run([bootstrap_path, *bootstrap_args], env=env, check=True)
[🔨]                       File "/usr/lib/python3.12/subprocess.py", line 571, in run
[🔨]                         raise CalledProcessError(retcode, process.args,
[🔨]                     subprocess.CalledProcessError: Command '[PosixPath('/tmp/tmpvdurduom/cmake-3.31.2/bootstrap'), '--prefix=/tmp/tmpvdurduom/cmake-install', '--no-qt-gui', '--no-debugger', '--parallel={parallel}', '--', '-DBUILD_TESTING=OFF', '-DBUILD_CursesDialog:BOOL=OFF']' returned non-zero exit status 7.
[🔨]                     [end of output]
[🔨]         
[🔨]                 note: This error originates from a subprocess, and is likely not a problem with pip.
[🔨]                 ERROR: Failed building wheel for cmake
[🔨]               Failed to build cmake
[🔨]               ERROR: Could not build wheels for cmake, which is required to install pyproject.toml-based projects
[🔨]               [end of output]
[🔨]         
[🔨]           note: This error originates from a subprocess, and is likely not a problem with pip.
[🔨]         error: subprocess-exited-with-error
[🔨]         
[🔨]         × pip subprocess to install backend dependencies did not run successfully.
[🔨]         │ exit code: 1
[🔨]         ╰─> See above for output.
[🔨]         
[🔨]         note: This error originates from a subprocess, and is likely not a problem with pip.
[🔨]         [end of output]
[🔨]     
[🔨]     note: This error originates from a subprocess, and is likely not a problem with pip.
[🔨]   error: subprocess-exited-with-error
[🔨]   
[🔨]   × pip subprocess to install build dependencies did not run successfully.
[🔨]   │ exit code: 1
[🔨]   ╰─> See above for output.
[🔨]   
[🔨]   note: This error originates from a subprocess, and is likely not a problem with pip.
[💥] Error 1 occurred in main shell [ at /home/leo/DATA/armbian/lib/functions/logging/runners.sh:211
    run_host_command_logged_raw() --> lib/functions/logging/runners.sh:211
        run_host_command_logged() --> lib/functions/logging/runners.sh:193
         prepare_python_and_pip() --> lib/functions/general/python-tools.sh:132
   aggregate_all_packages_python() --> lib/functions/configuration/aggregation.sh:47
             aggregate_packages() --> lib/functions/configuration/aggregation.sh:35
                do_with_logging() --> lib/functions/logging/section-logging.sh:81
   aggregate_packages_in_logging_section() --> lib/functions/configuration/aggregation.sh:23
       main_default_start_build() --> lib/functions/main/start-end.sh:25
          do_with_default_build() --> lib/functions/main/default-build.sh:41
         cli_standard_build_run() --> lib/functions/cli/cli-build.sh:25
        armbian_cli_run_command() --> lib/functions/cli/utils-cli.sh:136
                 cli_entrypoint() --> lib/functions/cli/entrypoint.sh:176
                           main() --> compile.sh:50
```

I got the impression that the native build is poorly supported.

pyavitz commented:
I've never tried to build native `RISCV` with armbian/build. Is it even supported?

The-going commented:
> I've never tried to build native `RISCV` with armbian/build. Is it even supported?

Swears at the C++ version. He wants 11 and 13 is installed.

I just tried to assemble it on the board.

Do not pay attention.
To be honest, I do not know how to fix all these dependencies for the build system.

The-going commented:
Build on QEMU KVM VM on x86_64 arch:
```
  ___ ___  _   ___ ___    __  __ _         _ 
 / __| _ \/_\ / __| __|__|  \/  (_)_ _  __| |
 \__ \  _/ _ \ (__| _|___| |\/| | | ' \/ _` |
 |___/_|/_/ \_\___|___|  |_|  |_|_|_||_\__,_|
                                             
 v25.02 rolling for BananaPi BPI-F3 running Armbian Linux 6.6.67-current-spacemit

 Packages:     Ubuntu stable (noble)
...
bananapif3:~# uname -r                                                                                                                      
6.6.67-current-spacemit
```

The-going commented:
What is the reason for this?
`declare -g KERNELBRANCH='commit:d3300881e2d524fbc4f9a8c7174df910f55ab3d0'`
I just didn't see this commit.
The branch has been rebased.
But the compilation is successful.

pyavitz commented:
> What is the reason for this? `declare -g KERNELBRANCH='commit:d3300881e2d524fbc4f9a8c7174df910f55ab3d0'` I just didn't see this commit. The branch has been rebased. But the compilation is successful.

I didn't want it to just pull willy-nilly, as things could break. So I'm having it pull last tested commit as they aren't using tags. Looks like they are working on the repo now and updated it to 6.6.68.

The-going commented:
I fixed the u-boot update feature. Now she checks all the files before writing.
And it won't write/overwrite if the file is already there.

And it now meets the requirements of shellfmt, shellcheck

pyavitz commented:
> I fixed the u-boot update feature. Now she checks all the files before writing. And it won't write/overwrite if the file is already there.
> 
> And it now meets the requirements of shellfmt, shellcheck

What you had in there before was busted, which is why I changed it.
It was overly complicated and why would the files not be there?

Anyway. if what you changed works. Kool.

pyavitz commented:
> I fixed the u-boot update feature. Now she checks all the files before writing. And it won't write/overwrite if the file is already there.
> 
> And it now meets the requirements of shellfmt, shellcheck

```
Skip  bootinfo_emmc.bin, it is equal to the existing one
# Write fw_dynamic.itb to /dev/mmcblk2boot0
266+1 records in
266+1 records out
# Write FSBL.bin to /dev/mmcblk2boot0
375+1 records in
375+1 records out
# Write u-boot.itb to /dev/mmcblk2boot0
4782+1 records in
4782+1 records out
```

This is what happens when using the armbian-config tool to transfer to eMMC. Its wrong and I also don't understand why we would want to skip anything?

```
echo 0 > /sys/block/mmcblk2boot0/force_ro
sleep .50
dd if="bootinfo_emmc.bin" of="/dev/mmcblk2boot0" bs=512 conv=notrunc
dd if="FSBL.bin" of="/dev/mmcblk2boot0" bs=512 seek=1 conv=notrunc
dd if="FSBL.bin" of="/dev/mmcblk2boot0" bs=512 seek=512 conv=notrunc
dd if="fw_dynamic.itb" of="/dev/mmcblk2" bs=512 seek=1280 conv=notrunc
dd if="u-boot.itb" of="/dev/mmcblk2" bs=512 seek=2048 conv=notrunc
sync
```

The-going commented:
> What you had in there before was busted, which is why I changed it.
> It was overly complicated and why would the files not be there?

This incorrect behavior has been fixed in https://github.com/armbian/build/pull/7554

> Anyway. if what you changed works. Kool.

There are two ways to check this.
Just build the u-boot package, install it on the platform and try to burn it only using armbian-install.

The second way is to see the details using a debugging script.:
```bash
#! /bin/bash

# dd if=$1/bootinfo_emmc.bin of=$device bs=512 conv=notrunc
# dd if=$1/FSBL.bin of=$device bs=512 seek=1 conv=notrunc
# dd if=$1/fw_dynamic.itb of=${2} bs=512 seek=1280 conv=notrunc
# dd if=$1/u-boot.itb of=${2} bs=512 seek=2048 conv=notrunc

write_uboot_platform() {
        local device N

        declare -A d
        d=(
                ["bootinfo_emmc.bin"]="0:$(du -b ${1}/bootinfo_emmc.bin | awk '{print $1}')"
                ["FSBL.bin"]="1:$(du -b ${1}/FSBL.bin | awk '{print $1}')"
                [fw_dynamic.itb]="1280:$(du -b ${1}/fw_dynamic.itb | awk '{print $1}')"
                [u-boot.itb]="2048:$(du -b ${1}/u-boot.itb | awk '{print $1}')"
        )

        device=${2}
        if [ -b ${2}boot0 ]; then
                device="${2}boot0"
        #       echo 0 > /sys/block/$(basename ${device})/force_ro
                echo "# sync"
        fi

        for f in "${!d[@]}"; do

                N=$((${d[$f]%:*} * 512))
                dd "if=${device}" bs=1 "skip=$N" "count=${d[$f]#*:}" conv=notrunc status=noxfer | xxd -l120
                echo ""
                xxd -l120 "${1}/${f}"
                echo "=========================================="
                if dd "if=${device}" bs=1 "skip=$N" "count=${d[$f]#*:}" \
                        conv=notrunc status=noxfer 2>/dev/null | cmp --quiet "${1}/${f}"
                then
                        echo "Skip  $f, it is equal to the existing one"
                else
                        echo "# Write =: N=$N d=${d[$f]} f=$f to ${device} seek=${d[$f]%:*}"
                        #dd if=$1/$f of=${device} bs=512 seek="${d[$f]%:*}" conv=notrunc,fsync status=noxfer
                fi
        done
}

write_uboot_platform "/usr/lib/linux-u-boot-legacy-bananapif3" "/dev/mmcblk0"

```
You must first install **_xxd_**

this script doesn't record anything, it just shows what it's checking.

The-going commented:
> This is what happens when using the armbian-config tool to transfer to eMMC. Its wrong and I also don't understand why we would want to skip anything?

`Skip  bootinfo_emmc.bin, it is equal to the existing one`

Why should we record the same thing that has already been recorded?
In any case, just undo the last commit if it doesn't suit you.

The-going commented:
I'm going on New Year's holidays and will be available after January 15th.
Happy New Year!

pyavitz commented:
> I'm going on New Year's holidays and will be available after January 15th. Happy New Year!

I'm just gonna REVERT until it can be sorted. I don't feel like fussing with it. The other function might not look like a coders dream, but it did work.

Enjoy and Happy New Year.

pyavitz commented:
This is beyond the scope of this PR. The builder says RISCV isn't supported when trying to use Trixie, when in fact Trixie does have RISCV support. This should be corrected at some point.

I'm done with this PR.
